### PR TITLE
Unet3d model bring up start: working ttnn/reference impl and passing …

### LINKED ISSUE #30858: [DRAFT] [Bounty $1500] 3D DenseUNet bring up using TTNN APIs
### 📝 Background

This bounty is for bringing up a [3D DenseUNet](https://arxiv.org/abs/1606.06650) segmentation model using TTNN APIs on Tenstorrent hardware (Wormhole or Blackhole). 3D DenseUNet combines the encoder-decoder structure of U-Net with densely connected blocks, which has been shown effective for volumetric (3D) medical image segmentation such as liver/tumor CT or brain MRI. Your goal is to enable a clean, correct 3D DenseUNet baseline and then optimize it for TT hardware throughput and latency.

### 🎯 What Success Looks Like

A successful submission will fulfill all requirements in the following stages. Payout is made after **all three stages** are completed.

### Stage 1 — Bring-Up

- Implement 3D DenseUNet using TTNN APIs (Python).
- Model runs on N150 or N300 without runtime errors.
- Produces valid 3D segmentations (e.g., volumetric mask outputs) on sample volumes.
- Output is verifiable vs. a PyTorch reference (per-case Dice / IoU on a small validation set).

### Stage 2 — Basic Optimizations

- Choose sharding/interleaving strategies suited for 3D conv/upsample/concat stages.
- Fuse common ops where possible (e.g., conv + norm + activation) via TT fused ops.
- Store intermediates in L1 when beneficial.
- Use recommended TTNN/tt-metal CNN flows.
- Use TT library of fused ops like [relu with conv](https://github.com/tenstorrent/tt-metal/pull/29236)
- Profile and remove redundant Tensor Manipulations (TMs) in the decode path (upsample + concat).

### Stage 3 — Deeper Optimization

- Scale up core parallelism for 3D convolutions and dense blocks.
- Explore block re-ordering / grouping to minimize TM and memory traffic.
- Document any advanced tuning or known issues.

### 🧭 Guidance & Starting Points

- Use the [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md) as your primary reference.
- Reference [Yolov4 implementation (PR 29157)](https://github.com/tenstorrent/tt-metal/pull/29157) for example end-to-end flows.
- Refer to [TT fused ops](https://github.com/tenstorrent/tt-metal/pull/29236) for opportunities to optimize.
- Ask for help or file issues if ops are missing in TTNN.

### 🔎 Possible Approaches

- Start from a PyTorch 3D DenseUNet (or 3D U-Net + Dense blocks) and port modules incrementally to TTNN.
- Validate each major block (dense block, transition, upsampling, concat) against a CPU/PyTorch reference.
- Experiment with chunked volumetric inference (sliding-window or patch-based) if memory is a constraint.
- Use TTNN profiling to identify heavy 3D conv and skip-concat hotspots; apply fusion and layout tweaks.
- Open a draft PR early to get feedback on your approach.

### 📚 Resources

- [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md)
- [CNN Bring-up & Optimization in TT-NN](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/CNNs/cnn_optimizations.md)
- [Yolov4 TTNN implementation PR](https://github.com/tenstorrent/tt-metal/pull/29157)
- [PyTorch 3D DenseUNet implementation](https://github.com/NguyenJus/pytorch-dense-unet-3d)
- [TT fused ops PR](https://github.com/tenstorrent/tt-metal/pull/29236)
- [Perf report header](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

### PR DESCRIPTION
**Pull request by mert-kurttutan**
### Ticket
Linked issue #30858

### Problem description
Model bring up for 3D DenseUNet

### What's changed
- Reference (pytorch) and ttnn implementation
- Optimization (e.g. fusion, deallocation of tensors) when not needed, sharding, model tracing
- Demo and validation using a public dataset
- Testing for perf and pcc

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable)
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml) job passes
- [ ] New/Existing tests provide coverage for changes

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212832436210770

### CODE REVIEW

**Comment by mert-kurttutan** [comment]
This is the start of my PR.
Regarding model implementations, since the model is 3D, it has some ops that is implemented natively, e.g. max_pool3d or upsample. For these, I implemented a temporary solution by using 2d version twice where you apply ops on along h,w dimensions and treat t dimension as batch. Then, you combine h w dimensions into a single dimension and treat t dimension as another dim. 
There seems to be 3 options:
- Use 2d version
- Use torch version as a fallback
- And/or open issue for the native 3d implementation of these ops.

**Comment by mert-kurttutan** [comment]
Another issue was that some of the ops might use some memory optimization. For instance max_pool2d seems to allocating the additional memory as big as the input, which forces me to use e.g. `64x64x64`x2 instead of `64x64x128`.
If this op could be done with low memory allocation (or even inplace), it would be great in terms of fitting model run inside L1

**Comment by mert-kurttutan** [comment]
Below, I attached the perf sheet and validation log that includes the validation result produced by validate script
[ops_perf_results_unet3d_2026_01_04_20_07_02.csv](https://github.com/user-attachments/files/24424246/ops_perf_results_unet3d_2026_01_04_20_07_02.csv)

[validation.log](https://github.com/user-attachments/files/24424253/validation.log)

**Comment by mert-kurttutan** [comment]
@cglagovichTT Could you help with the review ?

**Comment by yieldthought** [comment]
Codex review:

- [P2] Persist predictions to the HDF5 output file — [models/demos/unet_3d/demo/predictor.py:78](https://github.com/tenstorrent/tt-metal/blob/49c16a7d53bf7a88fd6c0cfc65a84f289fdf2e4f/models/demos/unet_3d/demo/predictor.py#L78)
  The inference loop writes predictions into a NumPy array returned by `_allocate_prediction_array`, but nothing ever creates a dataset in `h5_output_file` (and `_create_prediction_dataset` is never called). When running the validation/demo pipeline, the `.h5` file is created without the `output_dataset` dataset, so downstream consumers won’t find predictions. Consider creating the dataset in the HDF5 file (or calling `_create_prediction_dataset`) before writing patch outputs.

- [P2] Use full halo when computing test input resolution — [models/demos/unet_3d/runner/performant_runner_infra.py:69](https://github.com/tenstorrent/tt-metal/blob/49c16a7d53bf7a88fd6c0cfc65a84f289fdf2e4f/models/demos/unet_3d/runner/performant_runner_infra.py#L69)
  The test infra builds its input resolution as `patch_shape + halo_shape`, but the dataset pads halo on both sides (patch size grows by `2 * halo`). If a non-zero halo is configured (which the dataset warns is needed to avoid checkerboard artifacts), the runner will trace an input shape that’s smaller than actual inference patches, leading to shape mismatches or incorrect output. This should use `p + 2 * h` (or otherwise mirror the padded patch size).

**Comment by yieldthought** [comment]
Tested PR #34920 (head 49c16a7d53b) on IRD wormhole_b0 single device (yyzc-wh-01) on Jan 21, 2026.

Build:
- ./create_venv.sh --force
- ./build_metal.sh

Accuracy/PCC:
- pytest models/demos/unet_3d/tests/pcc/test_model.py -q -> 2 passed (PCC threshold 0.97)

Perf (single device):
- pytest models/demos/unet_3d/tests/perf/test_e2e_performant.py::test_unet3d_e2e -q -s -> PASSED
  - PCC=0.9946582865671103
  - avg iter time 0.033485s (FPS ~30)

Notes to reproduce:
- venv was missing pip; ran: python -m ensurepip --upgrade
- demo deps in venv: python -m pip install safetensors h5py tqdm scikit-image
- perf test required weights: python models/demos/unet_3d/demo/download_models.py (creates models/demos/unet_3d/data/models/confocal_boundary.safetensors)

Not run:
- 2-device mesh perf test (test_unet3d_e2e_dp)
- validation demo (requires dataset download)

**Review by yieldthought — APPROVED** [review]
I think this is looking good - do take a look at the codex comments. Approving to unblock!

**Comment by jberkowitzTT** [comment]
Hi @mert-kurttutan - once the CI check errors are resolved, this should be good to merge.

Please let me & @minaliuTT know once that's complete. Thank you for your hard work on this!

**Comment by mert-kurttutan** [comment]
@jberkowitzTT regarding errors, they seem to be unrelated to the code change in this PR. One of them is because of sfpi version (maybe related to my branch of ouf sync at the time, it could be fixed after I rebased it). The other one is related to perf csv file not having correct columns. I am not sure about the second one

**Comment by mert-kurttutan** [comment]
> I think this is looking good - do take a look at the codex comments. Approving to unblock!

Just commited to address it. Thanks for the review !!

**Comment by mert-kurttutan** [comment]
@jberkowitzTT The error regarding csv file is an internal error [1]. It essentially does not put the "branch line" when it checks out from detached head (which my PR request is, as is every PR request). The corresponding logic is [here](https://github.com/tenstorrent/tt-metal/blob/f1f5abc8598426186251ccb7b59fd8ca0a55b1c7/models/perf/perf_utils.py#L30). This missing line leads to early stop error for the next function.


1: The csv file is supposed to be an output of the following internal command
```shell
pytest -n auto --timeout 300 models/experimental/stable_diffusion_xl_base/tests/test_sdxl_op_unit_test_perf.py -k "_performance" -xv --tb=short
```

**Comment by dvartaniansTT** [comment]
> This is the start of my PR. Regarding model implementations, since the model is 3D, it has some ops that is implemented natively, e.g. max_pool3d or upsample. For these, I implemented a temporary solution by using 2d version twice where you apply ops on along h,w dimensions and treat t dimension as batch. Then, you combine h w dimensions into a single dimension and treat t dimension as another dim. There seems to be 3 options:
> 
> * Use 2d version
> * Use torch version as a fallback
> * And/or open issue for the native 3d implementation of these ops.

please include these details and the workaround in the model README.

**Review by dvartaniansTT — CHANGES_REQUESTED** [review]
please include a polished version of such details and workaround in the README file. 

"This is the start of my PR. Regarding model implementations, since the model is 3D, it has some ops that is implemented natively, e.g. max_pool3d or upsample. For these, I implemented a temporary solution by using 2d version twice where you apply ops on along h,w dimensions and treat t dimension as batch. Then, you combine h w dimensions into a single dimension and treat t dimension as another dim. There seems to be 3 options:

Use 2d version
Use torch version as a fallback
And/or open issue for the native 3d implementation of these ops."

**Comment by mert-kurttutan** [comment]
> > This is the start of my PR. Regarding model implementations, since the model is 3D, it has some ops that is implemented natively, e.g. max_pool3d or upsample. For these, I implemented a temporary solution by using 2d version twice where you apply ops on along h,w dimensions and treat t dimension as batch. Then, you combine h w dimensions into a single dimension and treat t dimension as another dim. There seems to be 3 options:
> > 
> > * Use 2d version
> > * Use torch version as a fallback
> > * And/or open issue for the native 3d implementation of these ops.
> 
> please include these details and the workaround in the model README.

@dvartaniansTT Thanks for the review, just added it

**Comment by minaliuTT** [comment]
> > > This is the start of my PR. Regarding model implementations, since the model is 3D, it has some ops that is implemented natively, e.g. max_pool3d or upsample. For these, I implemented a temporary solution by using 2d version twice where you apply ops on along h,w dimensions and treat t dimension as batch. Then, you combine h w dimensions into a single dimension and treat t dimension as another dim. There seems to be 3 options:
> > > 
> > > * Use 2d version
> > > * Use torch version as a fallback
> > > * And/or open issue for the native 3d implementation of these ops.
> > 
> > 
> > please include these details and the workaround in the model README.
> 
> @dvartaniansTT Thanks for the review, just added it

@dvartaniansTT The contributor has added the README details and workaround you requested. When you have a chance, could you take another look and re-approve if it looks good? Thanks!

**Review by dvartaniansTT — APPROVED** [review]

**Comment by fvranicTT** [comment]
@mert-kurttutan could you please rebase against the latest main? AI Assistant error should go away.

**Review by dvartaniansTT — APPROVED** [review]
